### ISSUE TITLE
[SEC][S0] SafeTensors: an unknown wire dtype skips all strict offset validation

### ISSUE THREAD

**Issue by kekzl** [kekzl/imp#1603]
## Fact
- A SafeTensors tensor whose `dtype` string is not one of the 14 known wire types skips `validate_tensor_offsets()` entirely, so `offset_start` is never validated at all and the only remaining check adds two attacker-controlled `uint64`s.
- `src/model/safetensors_loader.cpp:473`
- `src/model/safetensors_loader.cpp:477`
- `src/model/safetensors_loader.cpp:496`

## Evidence
```
src/model/safetensors_loader.cpp:472
        size_t wire_bytes = safetensors_wire_dtype_bytes(dtype_val->str_val);
        if (wire_bytes == 0) {
            // Unknown wire dtype — skip strict validation but keep the soft
            // OOB check below. safetensors_dtype()'s WARN already fires for
            // unknown wire types when the tensor is actually emitted.
            if (tensor_data_offset + offset_end > file_size) {
                n_dropped_offset_validation++;
                warn_drop(tensor_name.c_str(), "offset_end past EOF (unknown wire dtype, lenient check)");
                continue;
            }
        } else {
```
```
src/model/safetensors_loader.cpp:496
        void* tensor_ptr = tensor_data_base + offset_start;
        Tensor t(tensor_ptr, dtype, ndim, shape, /*on_device=*/false);
```
- `offset_start` appears nowhere between line 467 (where it is read from the file) and line 496 (where it becomes a raw pointer) on this branch. `validate_tensor_offsets` is the only function that checks it, and it is called only in the `else` branch at :488.
- The same file rejects exactly this addition pattern 430 lines earlier, and says why:
```
src/model/safetensors_loader.cpp:39
    // Overflow-safe form: declared_header_size > (file_size - 8) cannot
    // overflow because file_size - 8 is a valid subtraction (file_size >= 8).
    // The naive 8 + declared_header_size > file_size would wrap to a small
    // value when declared_header_size approaches UINT64_MAX, bypassing the
    // check.
    if (declared_header_size > file_size - 8) {
```
- The branch is reachable: an unknown dtype string is not a load failure, it is a WARN plus an F32 default.
```
src/model/safetensors_loader.cpp:145
    IMP_LOG_WARN("Unknown SafeTensors dtype '%s', defaulting to FP32", s.c_str());
    return QType::F32;
```
- The emitted tensor's byte count is then recomputed from the shape, not from the file:
```
src/core/tensor.cpp:37
size_t Tensor::nbytes() const {
    int64_t n = numel();
    ...
    return static_cast<size_t>(n) * qtype_elem_bytes(qtype);
}
```
```
src/model/weight_upload.cu:696
    int64_t n_elem = weight.numel();
    const float* src = static_cast<const float*>(weight.data);
    std::vector<uint16_t> h_fp16(static_cast<size_t>(n_elem));

    for (int64_t i = 0; i < n_elem; ++i) {
        h_fp16[i] = float_to_fp16(src[i]);
    }
```

## Impact
- A checkpoint with `"dtype": "F8_E8M0"` (or any unrecognised string) plus `data_offsets: [<large>, 0]` yields a `Tensor` whose base pointer is `mmap_base + 8 + header_size + offset_start` with `offset_start` unbounded — an arbitrary out-of-mapping read address.
- `tensor_data_offset + offset_end` wraps for `offset_end >= 2^64 - tensor_data_offset`, so even the lenient check is defeatable without touching `offset_start`.
- Host-side out-of-bounds read of `numel() * 4` bytes at `weight_upload.cu:700`, copied into device memory and thereby into logits: SIGSEGV, or process-memory disclosure through generated tokens.
- Blast radius: every SafeTensors load path — `--model <dir>`, `--models-dir` auto-load on first request, and any HuggingFace repo id resolved through `resolve_model_auto` / `src/model/hf_hub.cpp`. GGUF is unaffected; its equivalent check saturates instead (`src/model/gguf_parse.cpp:440-459`).

## Repro
```bash
# Static proof that offset_start is unchecked on this branch:
cd /home/kekz/github.com/kekzl/imp
sed -n '467,498p' src/model/safetensors_loader.cpp | grep -n "offset_start"
# -> line 1 (read from file) and line 30 (used as a pointer). Nothing in between.

# Static proof the strict validator is the only checker of offset_start:
grep -rn "offset_start" src/model/safetensors_loader.cpp

# NOT RUN (crafted-file repro, needs a writable scratch dir and a load):
#   build a model.safetensors with header {"w":{"dtype":"F8_E8M0","shape":[4,4],
#   "data_offsets":[18446744073709551000,0]}} and run imp-cli --model <dir>
```

## Fix direction
- Call `validate_tensor_offsets()` on every tensor, including unknown dtypes; pass `expected_nbytes = 0` and add a "size unknown" mode that still enforces `offset_start <= offset_end <= file_size - tensor_data_offset`.
- Or refuse an unknown `dtype` string outright at load, consistent with `src/model/CLAUDE.md`'s stated invariant "a checkpoint this build cannot serve must be refused at load".
- Never write `a + b > limit` on file-derived `uint64` in this file; the subtraction form is already established at `:39-46`.

## Acceptance criteria
- [ ] Loading a SafeTensors file with an unknown `dtype` and `offset_start > file_size` returns a load error, not a `Tensor`.
- [ ] Loading a SafeTensors file with `offset_end = UINT64_MAX - 8` is rejected (no wrap).
- [ ] `grep -n "tensor_data_offset + offset_end" src/model/safetensors_loader.cpp` returns nothing.
- [ ] A test in `tests/test_safetensors_loader.cpp` drives the real `load_shard` with both crafted headers, in the `unit` ctest label so CI runs it.

## Unverified
- Whether a wrapped `offset_start` faults immediately or reads a neighbouring mapping is host-dependent; the OOB is proven statically, the observable outcome is not. Settle with the crafted-file command above under `make asan`.

**Comment by kekzl** [comment]
Correction accepted, verified here, and the severity does not move.

**Impact bullet 2 is not reachable over the JSON path.** `data_offsets` arrives through `JsonValue::as_int()`:
```
src/model/json_util.h:29:    int64_t as_int() const { return static_cast<int64_t>(num_val); }
```
That is a `double -> int64_t` conversion, so a header value near 2^64 saturates to `INT64_MIN` before it
ever reaches `tensor_data_offset + offset_end`, and the addition cannot wrap. The repro header quoted in
the body (`data_offsets: [18446744073709551000, 0]`) collapses the same way. That bullet is withdrawn.

**Impact bullet 1 stands and is the defect:** on the unknown-wire-dtype branch, `offset_start` is never
examined at all - the lenient check reads only `offset_end` - so the tensor is emitted with a pointer far
outside the mapping. That is an out-of-bounds read from a model file, so **S0 stays**.

Struck bullet 2 rather than the issue. Anyone reading this later: the useful general lesson is that every
numeric field in the SafeTensors header passes through a `double`, which bounds what a hostile value can
be and is worth checking before arguing an overflow anywhere else in that parser.

**Comment by kekzl** [comment]
Fixed in #1660 (squashed to e86420b1 on main). The PR body's single `Closes` line with a comma list only ever links the first number, so these are closed by hand.
